src/lcm/LifeCycleStates.cc

```diff
@@ -2496,3 +2673,500 @@ void LifeCycleManager::trigger_resize_failure(int vid)
+
+void LifeCycleManager::trigger_disk_restore_success(int vid)
+{
+    trigger([this, vid] {
+        Template vm_quotas_snp;
+        vector<Template *> ds_quotas_snp;
+        int uid = -1, gid = -1;
+
+        if ( auto vm = vmpool->get(vid) )
+        {
+            VirtualMachine::LcmState state = vm->get_lcm_state();
+
+            if (state == VirtualMachine::RESTORE)
+            {
+                uid = vm->get_uid();
+                gid = vm->get_gid();
+                vm->delete_snapshots(vm_quotas_snp);
+                vm->delete_non_persistent_disk_snapshots(vm_quotas_snp, ds_quotas_snp);
+
+                vm->set_state(VirtualMachine::POWEROFF);
+                vm->log("LCM", Log::INFO, "VM restore operation completed.");
+            }
+            else
+            {
+                vm->log("LCM", Log::ERROR, "VM restore success, VM in a wrong state");
+                return;
+            }
+
+            vmpool->update(vm.get());
+        }
+
+        if (!vm_quotas_snp.empty())
+        {
+            Quotas::quota_del(Quotas::VM, uid, gid, &vm_quotas_snp);
+        }
+
+        if ( !ds_quotas_snp.empty() )
+        {
+            Quotas::ds_del_recreate(uid, gid, ds_quotas_snp);
+        }
+
+    });
+}
+
+/* -------------------------------------------------------------------------- */
+/* -------------------------------------------------------------------------- */
+
+void LifeCycleManager::trigger_disk_restore_failure(int vid)
+{
+    trigger([this, vid] {
+        if ( auto vm = vmpool->get(vid) )
+        {
+            VirtualMachine::LcmState lcm_state = vm->get_lcm_state();
+
+            if (lcm_state == VirtualMachine::RESTORE)
+            {
+                vm->set_state(VirtualMachine::POWEROFF);
+                vm->log("LCM", Log::INFO, "VM restore operation fails");
+            }
+            else
+            {
+                vm->log("LCM", Log::ERROR,
+                        "restore fails, VM in a wrong state: " + vm->state_str());
+                return;
+            }
+
+            vmpool->update(vm.get());
+        }
+    });
+}
+
+/* -------------------------------------------------------------------------- */
+/* -------------------------------------------------------------------------- */
+
+static int create_backup_image(VirtualMachine * vm, string& msg)
+{
+    Nebula& nd = Nebula::instance();
+
+    auto ipool  = nd.get_ipool();
+    auto dspool = nd.get_dspool();
+
+    int i_id;
+    string error_str;
+
+    auto& backups = vm->backups();
+
+    /* ------------------------------------------------------------------ */
+    /* Get datastore backup information                                   */
+    /* ------------------------------------------------------------------ */
+    string ds_name, ds_mad, ds_data;
+
+    Datastore::DatastoreType ds_type;
+    Image::DiskType          ds_dtype;
+
+    int ds_id = backups.last_datastore_id();
+
+    if (auto ds = dspool->get_ro(ds_id))
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

This snippet may contain an accidental use of incorrect operation. Make sure that the operation being performed is in fact what you intended.